### ISSUE TITLE
git lfs locks --local data improvement suggestions

### ISSUE THREAD

**Issue by mstrap** [git-lfs/git-lfs#3107]
We would like to implement Git LFS locks functionality for our GUI client. One basic feature would be to display "known" locks. I was hoping to find this information using `git lfs locks --local`, but unfortunately this information is inappropriate or not available at all.

**Problem**

(1) `git lfs locks --local` will report for every *own* lock, the last known state after having invoked `git lfs lock/unlock <file>`. If in the meanwhile locks in the repository have changed (because they were forcefully unlocked or the lock was released from a different clone), this information becomes outdated and it will *remain outdated* even after invoking `git lfs locks` to query updated locks information.

(2) `git lfs locks --local` does not report locks of other users.

**Suggested solution**

To follow the distributed, offline-capable approach of Git in general, `git lfs locks --local` could mirror the last known information queried by `git lfs locks`. 

**Why solve in Git LFS code?**

(1) This would make Git LFS binaries somewhat more robust in case of offline usage and allow the user to still see last known locks information, without requiring a connection to the server.

(2) It will provide an API for (GUI) clients which may want to enrich `git status` information by Git LFS locks information. As `git status` information is purely local, it will be convenient to have the additional locks information be present locally, too, to report the combined result in a single run (without requiring a connection to the server). If Git LFS itself won't provide this information, this would probably mean that interested clients will have to implement their own caches (which will be harder to implement from outside of Git LFS code and which are probably more likely to become inconsistent).

**Comment by ttaylorr** [comment]
Hi @mstrap, thanks for opening this.

> **Suggested solution**
> 
> To follow the distributed, offline-capable approach of Git in general, `git lfs locks --local` could mirror the last known information queried by `git lfs locks`.

I think that this is a great idea, and we could definitely support it without too much craziness :-).

I am thinking that after every `git lfs locks` that contacts the remote server, we could update a file in `.git/lfs/cache/locks` (the `cache` sub-tree does not yet exist or serve any purpose) and hence read from that instead of the current use of `--local`.

`.git/lfs/cache/locks` could store the JSON response that we got from the remote verbatim, and would be updated when `git lfs locks` is issued. @PastelMobileSuit & @larsxschneider: what do you think of this? Would one of you be interested in implementing it?

OTOH, I don't think we can assign this functionality to `--local` before v3.0.0, since it has other behavior that we cannot break the backwards compatibility of between v2.x.y's. That said, we could introduce a `--cached` and mark `--local` as "behaviorally deprecated", and note that it will behave as `s/--local/--cached/g` in future versions.

What do you think?

**Comment by mstrap** [comment]
Thanks, @ttaylorr, your proposal sounds great to me and will work for us. Having raw information available in verbatim format is definitely a good idea.

**Comment by ttaylorr** [comment]
> TLDR: `git lfs locks --cached` makes total sense to me. I wouldn't retire `git lfs locks --local` though.

That seems sensible to me, and I think we should do that.

**Comment by mstrap** [comment]
I have just pieced together and pushed some code which is addressing writing of the cache file (on top of that the `--cached` option be implemented). Code is basically doing what it should. I haven't added any tests for now, but will write them once I get some feedback whether my approach is OK.

Btw, I'm new to the Git-LFS sources and even newer to Golang, so please be patient with me :) I'm happy to learn ...

**Comment by ttaylorr** [comment]
> I have just pieced together and pushed some code which is addressing
> writing of the cache file (on top of that the `--cached` option be
> implemented).

Hooray! Thank you for taking the time to work on this.

I would be more than happy to look at your code, but I think that the
best way to do so would be to communicate over a pull request. If you
haven't already, could you please open one against Git LFS's copy?
Thanks.
